mender-update/main.cpp

```diff
@@ -83,5 +84,45 @@ int main() {
 		std::cout << "Got (un)expected error: '" << ex_s_err.error().message << "'" << std::endl;
 	}
 
+	class PrintBody : public io::Writer {
+	public:
+		expected::ExpectedSize Write(
+			vector<uint8_t>::const_iterator start, vector<uint8_t>::const_iterator end) {
+			cout << string {start, end} << flush;
+			return end - start;
+		}
+		~PrintBody() {
+			cout << "Done!\n";
+		}
+	};
+	auto header_handler = [](mender::http::ExpectedHttpResponsePtr result) {
+		if (!result) {
+			log::Error(result.error().String());
+			return;
+		}
+		result.value()->SetBodyWriter(make_shared<PrintBody>());
```

Comment:
The `Response` takes shared ownership of the writer, which survives this function call, so we cannot instantiate it on the stack. It would be gone after this function returns, but it's still being written to asynchronously.

While a `move` constructor would be possible, it is not desirable here because the caller should be allowed to keep a reference to the object as well. We don't use that capability here, but if we had, the ownership needs to be shared in order to refer to the same object. So this style is more like a Golang pointer to object.

Reply:
Could we not have both? :thinking:

Reply:
I don't think so, the ownership semantics would be confusing if we had both, and we are not really gaining anything apart from avoiding one single allocation.